qiskit_experiments/curve_analysis/composite_curve_analysis.py

```diff
@@ -0,0 +1,452 @@
+        override :meth:`CompositeCurveAnalysis._create_curve_data` in subclass.
+
+    :class:`.CompositeCurveAnalysis` may override following methods.
+
+    .. rubric:: _evaluate_quality
+
+    This method evaluates the quality of the composite fit based on
+    the all analysis outcomes.
+    This returns "good" when all fit outcomes are evaluated as "good",
+    otherwise it returns "bad".
+
+    .. rubric:: _create_curve_data
+
+    This method is passed all the group fit outcomes and can return a list of
+    new values to be stored in the analysis results.
+
+    """
+
+    # TODO move example to curve analysis user guide. This is too specific to be written in
+    #  the API documentation of general framework.
+
+    def __init__(
+        self,
+        analyses: List[BaseCurveAnalysis],
+        name: Optional[str] = None,
+    ):
+        super().__init__()
+
+        self._analyses = analyses
+        self._name = name or self.__class__.__name__
+
+    @property
+    def name(self) -> str:
+        """Return name of this analysis."""
+        return self._name
+
+    @property
+    def parameters(self) -> List[str]:
+        """Return parameters of this curve analysis."""
+        unite_params = []
+        for analysis in self._analyses:
+            # Respect ordering of parameters
+            for name in analysis.parameters:
+                if name not in unite_params:
+                    unite_params.append(name)
+        return unite_params
+
+    @property
+    def models(self) -> Dict[str, lmfit.Model]:
+        """Return fit models."""
+        models = {}
+        for analysis in self._analyses:
+            models[analysis.name] = analysis.models
+        return models
+
+    def analyses(self, index: Optional[Union[str, int]] = None) -> BaseCurveAnalysis:
+        """Return curve analysis instance.
+
+        Args:
+            index: Name of group or numerical index.
+
+        Returns:
+            Curve analysis instance.
+        """
+        if index is None:
+            return self._analyses
+        if isinstance(index, str):
+            group_names = [analysis.name for analysis in self._analyses]
+            num_index = group_names.index(index)
+            return self._analyses[num_index]
+        return self._analyses[index]
+
+    def _generate_fit_guesses(
+        self,
+        user_opt: FitOptions,
+        curve_data: CurveData,  # pylint: disable=unused-argument
+    ) -> Union[FitOptions, List[FitOptions]]:
+        """Create algorithmic guess with analysis options and curve data.
+
+        Args:
+            user_opt: Fit options filled with user provided guess and bounds.
+            curve_data: Formatted data collection to fit.
+
+        Returns:
+            List of fit options that are passed to the fitter function.
+        """
+        # This method is delegated to self.analyses
+        return NotImplemented
```

Comment:
I think all these `NotImplemented` methods were pushed down from `CurveAnalysis` to `BaseCurveAnalysis` for the sake of `CompositeCurveAnalysis`. If we like how `CompositeCurveAnalysis` works now, should we consider lifting them back up?

Reply:
Or maybe `CompositeCurveAnalysis` just should inherit from `BaseAnalysis`? Is it using any code from `BaseCurveAnalysis`?

Reply:
Implementation of several methods (`_format_data`, `_run_data_processing`, `_run_curve_fit`) are moved to `CurveAnalysis`, and `_generate_fit_guesses` is removed from `BaseCurveAnalysis` because this is called inside `_run_curve_fit`. By replacing `CurveData` with data frame, we can implement these methods on `CompositeAnalysis` rather than implementing everything in `_run_analysis` as in this PR. This allows users to easily customize the behavior of data processing, formatting, and analysis without entirely overriding the `_run_analysis` method.

Please check https://github.com/Qiskit/qiskit-experiments/pull/824/commits/d3a5de4cedcd2918d0fee8fcd897ee2ffa5bab71

Reply:
How do you imagine `_format_data`, `_run_data_processing`, and `_run_curve_fit` being implemented in `CompositeCurveAnalysis` with the updated `CurveData` dataframe? Will you remove the loop over analyses from `_run_analysis` and have each smaller method do its own loop over groups? What is a scenario for which a user might want to override one of these methods? Most of the real logic would be in the individual `analyses`. I think the user would be more likely to override the methods of the classes of the `analyses`.

Reply:
Make sense. These methods are removed in https://github.com/Qiskit/qiskit-experiments/pull/824/commits/9b97df6acabe3443d31204495033eeda67a6aa7b. Now composite curve analysis is a direct subclass of `BaseAnalysis`.